hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/MemberPruningTest.java

```diff
@@ -0,0 +1,273 @@
+        instance().getMap(mapName).set(key, Integer.MAX_VALUE);
+
+        assertRowsAnyOrder(instance(),
+                "SELECT this FROM " + mapName + " WHERE stringField = ? AND longField = ?",
+                asList(key.getStringField(), key.getLongField()),
+                Collections.singletonList(new Row(Integer.MAX_VALUE)));
+
+        assertInvokedOnlyOnMembers(instance(), partitionOwnerMember);
+    }
+
+    @Test
+    public void testSelectWithUnionAll() {
+        configureMapWithAttributes(mapName, "stringField");
+
+        String otherMapName = "test_other_" + randomName();
+        createMapping(otherMapName, KeyPojo.class, Integer.class);
+        configureMapWithAttributes(otherMapName, "stringField");
+
+        populateMap();
+        KeyPojo otherKey = new KeyPojo(
+                generateKeyOwnedBy(partitionOwnerMember),
+                RAND.nextInt(),
+                Long.MAX_VALUE
+        );
+        instance().getMap(otherMapName).set(otherKey, Integer.MIN_VALUE);
+
+        assertRowsAnyOrder(instance(),
+                "(SELECT this FROM " + mapName + " WHERE stringField = ?) "
+                        + "UNION ALL "
+                        + "(SELECT this FROM " + otherMapName + " WHERE stringField = ?)",
+                List.of(key.getStringField(), otherKey.getStringField()),
+                List.of(new Row(Integer.MAX_VALUE), new Row(Integer.MIN_VALUE)));
+
+        assertInvokedOnlyOnMembers(instance(), partitionOwnerMember);
+    }
+
+    @Test
+    public void testSelectWithConjunctionAndArbitraryFilter() {
+        configureMapWithAttributes(mapName, "stringField");
+
+        populateMap();
+        assertRowsAnyOrder(instance(),
+                "SELECT this FROM " + mapName + " WHERE stringField = ? AND 1 = 1",
+                Collections.singletonList(key.getStringField()),
+                Collections.singletonList(new Row(Integer.MAX_VALUE)));
+
+        assertInvokedOnlyOnMembers(instance(), partitionOwnerMember);
+    }
+
+    @After
+    public void teardown() {
+        instance().getMap(mapName).destroy();
+    }
+
+    private HazelcastInstance getRandomMember() {
+        return instances()[RAND.nextInt(MEMBER_COUNT)];
+    }
+
+    private void populateMap() {
+        partitionOwnerMember = getRandomMember();
+        key = new KeyPojo(
+                generateKeyOwnedBy(partitionOwnerMember),
+                RAND.nextInt(),
+                Long.MAX_VALUE
+        );
+        instance().getMap(mapName).set(key, Integer.MAX_VALUE);
+    }
+
+    private void configureMapWithAttributes(String mapName, String... attributes) {
+        MapConfig mc = new MapConfig(mapName);
+        for (String attribute : attributes) {
+            mc.getPartitioningAttributeConfigs()
+                    .add(new PartitioningAttributeConfig(attribute));
+        }
+        instance().getConfig().addMapConfig(mc);
+    }
+
+    private void assertInvokedOnlyOnMembers(HazelcastInstance... members) {
```

Comment:
such methods and listener are in `SqlEndToEndTestSupport`. you could use it as base class and move `assertInvokedOnlyOnMembers` there